clang/lib/Sema/SemaBounds.cpp

```diff
@@ -4017,11 +4023,59 @@ namespace {
         << ObservedBounds << ObservedBounds->getSourceRange();
     }
 
+    // BlameAssignmentWithinStmt prints a diagnostic message that highlights the
+    // assignment expression in St that causes V's observed bounds to be unknown
+    // or not provably valid.  If St is a DeclStmt, St itself and V are
+    // highlighted.  BlameAssignmentWithinStmt returns the source location of
+    // the blamed assignment.
+    SourceLocation BlameAssignmentWithinStmt(Stmt *St, const VarDecl *V,
+                                             CheckingState State,
+                                             unsigned DiagId) const {
+      SourceRange SrcRange = St->getSourceRange();
+      auto BDCType = Sema::BoundsDeclarationCheck::BDC_Statement;
+
+      // For a declaration, show the diagnostic message that starts at the
+      // location of v rather than the beginning of St and return.  If the
+      // message starts at the beginning of a declaration T v = e, then extra
+      // diagnostics may be emitted for T.
+      SourceLocation Loc = St->getBeginLoc();
+      if (isa<DeclStmt>(St)) {
+        Loc = V->getLocation();
+        BDCType = Sema::BoundsDeclarationCheck::BDC_Initialization;
+        S.Diag(Loc, DiagId) << BDCType << V << SrcRange << SrcRange;
+        return Loc;
+      }
+
+      // If not a declaration, find the assignment (if it exists) in St to blame
+      // for the error or warning.
+      auto It = State.BlameAssignments.find(V);
+      if (It != State.BlameAssignments.end()) {
+        Expr *BlameExpr = It->second;
+        Loc = BlameExpr->getBeginLoc();
+        SrcRange = BlameExpr->getSourceRange();
+        
+        // Choose the type of assignment E to show in the diagnostic messages:
+        // assignment (=), decrement (--) or increment (++).
+        if (UnaryOperator *UO = dyn_cast<UnaryOperator>(BlameExpr)) {
+          if (UO->isIncrementOp())
+            BDCType = Sema::BoundsDeclarationCheck::BDC_Increment;
+          else if (UO->isDecrementOp())
```

Comment:
Are there plans to handle other unary operators like `unary minus`, `deref`, `address-of`, etc?

Reply:
No. Here `UO` must be a self-modifying unary operator, and I can only think of `++` and `--`.

I can also put an assert here.